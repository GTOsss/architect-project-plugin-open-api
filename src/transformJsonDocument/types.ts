/**
 * Все возможные значения поля OpenAPIV3.SchemaObject.type кроме array и object.
 * */
type OpenAPIPrimitiveType = 'boolean' | 'number' | 'string' | 'integer';

/**
 * Map-объект для трансформации типов в нужный формат.
 * @example:
 * const typesMap = {
 *     integer: 'number',
 *     string: 'string',
 *     boolean: 'boolean',
 *     anyObject: 'AnyObject', // если в scheme только type: object
 *     '*': 'any', // any если не удалось определить тип
 *   };
 * */
export type TypesMap = Record<OpenAPIPrimitiveType | 'anyObject' | '*', string>;

export type Config = {
  typesMap: TypesMap;
};
