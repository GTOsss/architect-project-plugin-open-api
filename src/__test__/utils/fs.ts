import { parse } from 'yaml';
import { readFileSync } from 'fs-extra';

/**
 * Import yaml file as object
 * @example
 * readYaml(resolve(__dirname, '../../mock/openapi.mock.yaml'))
 * */
export const readYaml = (path: string) => {
  return parse(readFileSync(path, 'utf8'));
};
