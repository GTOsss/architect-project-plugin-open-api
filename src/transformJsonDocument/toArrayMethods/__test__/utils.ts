import { OpenAPIV3 } from 'openapi-types';
import { toArrayMethods } from '../toArrayMethods';
import { Method } from '../toArrayMethods.types';
import { TypesMap } from '../../types';

const typesMap: TypesMap = {
  integer: 'number',
  number: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

/**
 * Находит объект указанного метода.
 * @example findMethod(HttpMethods.GET, 'users/{id}', methods);
 * */
export const findMethod = (method: OpenAPIV3.HttpMethods, path: string, methods: Method[]): Method => {
  return methods.find((currentMethod) => currentMethod.method === method && currentMethod.path === path);
};

/**
 * Вызывает getMethodsAsArray и находит в полученном массиве указанный метод.
 * */
export const testFor = (method: OpenAPIV3.HttpMethods, path: string, jsonDocument: OpenAPIV3.Document): Method => {
  const methods = toArrayMethods(jsonDocument, { typesMap });
  return findMethod(method, path, methods);
};
