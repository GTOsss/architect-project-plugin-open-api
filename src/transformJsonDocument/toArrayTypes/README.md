Рассмотрим простой пример описания OpenAPI схемы и того как ее будет обрабатывать функция `toArrayTypes`.
Есть схемы:
```yaml
components:
  schemas:
    User:
      type: object
      properties:
        id:               
          type: integer   
          format: int64   
        tags:             
          type: array     
          items:          
            type: string  
        address:
          $ref: '#/components/schemas/Address'
    AnotherSchema:
      type: object
    # <...>
```
В JSON формате это выглядит так:
```js
const json = {
  "components": {
    "schemas": { // ← Мап-объект по этому нужен:
      /* 
        Object.entries() => [[schemaName, schemaData], ...], где:
          schemaName - Имя Схемы (так же имя будущего Type)
          schemaData - Объект схемы (объект с информацией для создания будущего Type)
      */
      
      "User": { // ← schemaName: "User"
        // ↓ schemaData ↓  
        "type": "object", // 
        "properties": { 
          "id": {
            "type": "integer",
            "format": "int64"
          },
          "tags": {
            "type": "array",
            "items": {
              "type": "string"
            }
          },
          "address": {
            "$ref": "#/components/schemas/Address"
          }
        }
        // ↑ schemaData ↑ 
      },
      
      "AnotherSchema" : { // ← schemaName: "AnotherSchema"
        // ↓ schemaData ↓
        type: "object"
        /* ... */
      }
    }
  }
}
```

## [toArrayTypes()](toArrayTypes.ts) → [transformSchema()](../utils/transformSchema/transformSchema.ts) 
Метод `toArrayTypes` проходит по мап-объекту `components/schemas` и каждую схему передает в метод 
[transformSchema()](../utils/transformSchema/transformSchema.ts).
```yaml
components:
  schemas:
    # transformSchema(schemaData: { весь объект ↓ } )
    User: 
      type: object
      properties:
        id:               
          type: integer   
          format: int64   
        tags:             
          type: array     
          items:          
            type: string  
        address:
          $ref: '#/components/schemas/Address'
    # transformSchema(schemaData: { весь объект ↓ } )
    AnotherSchema: 
      type: object
    # <...>
``` 

transformSchema()

## [transformSchema()](../utils/transformSchema/transformSchema.ts) → [createTypeObject()](../utils/transformSchema/transformSchema.ts) 

Описание метода: `transformSchema()`. Рассмотрим структуру User схемы и того как ее обрабатывает метод `transformSchema()`.
```yaml
    User:
      type: object
      # Запускается цикл для обхода данного объекта. В коде используется:
      # Object.entries(schema.properties).reduce((acc, [*fieldName*, *fieldSchema*] => {...}, [])
      properties:
        id:               # *fieldName* = 'id'
                          # *fieldSchema* = {
          type: integer   #   type: 'integer',
          format: int64   #   format: 'int64',
                          # }
                        
        tags:             # *fiendName* = 'tags'
                          # *fieldSchema* = {
          type: array     #   type: 'array'
          items:          #   items: {
            type: string  #      type: 'string'
                          #   }
                          # }
        
        address:                                 # *fieldName* = 'address'
                                                 # *fieldSchema* = {
          $ref: '#/components/schemas/Address'   #    $ref: '#/components/schemas/Address'
                                                 # }
```

