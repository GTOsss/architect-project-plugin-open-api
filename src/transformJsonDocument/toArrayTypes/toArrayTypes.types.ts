export type GeneratedType = {
  /** Имя типа */
  typeName: string;
  /** Значение типа. То что пишется после знака "="
   * @example
   * type TypeName = { a: string } // typeValue → "{ a: string }"  */
  typeValue: string;
};
