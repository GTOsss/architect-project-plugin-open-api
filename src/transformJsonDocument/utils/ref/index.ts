export * from './resolveRef';
export * from './getSchemaNameByRef';
export * from './parseRefPath';
