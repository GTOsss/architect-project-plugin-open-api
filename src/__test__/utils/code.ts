import { existsSync } from 'fs';
import { ESLint } from 'eslint';

type ApplyEslintAutoFixParams = {
  code: string;
  eslintConfigPath: string;
  filePath: string;
};

/**
 * Apply autofix for input code
 * */
export const applyEslintAutofix = async ({
  code,
  eslintConfigPath,
  filePath,
}: ApplyEslintAutoFixParams): Promise<string> => {
  try {
    if (!existsSync(eslintConfigPath)) {
      throw new Error(`ESLint configuration file not found by path: ${eslintConfigPath}`);
    }

    const eslint = new ESLint({
      fix: true,
      overrideConfigFile: eslintConfigPath,
    });

    const results = await eslint.lintText(code, { filePath });

    await ESLint.outputFixes(results);

    return results[0]?.output || code;
  } catch (error) {
    console.error('Error applying eslint autofix:', error);
    return code;
  }
};

export const applyEslint = async (code: string) =>
  await applyEslintAutofix({ code, filePath: 'temp/forEslintAutofix.ts', eslintConfigPath: 'eslint.config.mjs' });
