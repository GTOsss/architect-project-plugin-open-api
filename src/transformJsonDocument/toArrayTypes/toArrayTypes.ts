import { OpenAPIV3 } from 'openapi-types';
import { Config } from '../types';
import { transformSchema } from '../utils/transformSchema';
import { genCode } from '#/templates';

/**
 * Подробное описание логики функции в ./README.md
 * */
export const toArrayTypes = (jsonDocument: OpenAPIV3.Document, config: Config): string[] => {
  const schemas = jsonDocument.components.schemas;
  return Object.entries(schemas).reduce((acc, [schemaName, schemaData]) => {
    const typeValue = transformSchema(jsonDocument, schemaData, config);
    const schemaType = genCode('export type ~name~ = ~typeValue~', { name: schemaName, typeValue });
    acc.push(schemaType);
    return acc;
  }, [] as string[]);
};
