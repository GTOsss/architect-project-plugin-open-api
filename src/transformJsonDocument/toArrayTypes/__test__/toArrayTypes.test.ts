import { readYaml } from '../../../__test__/utils/fs';
import { resolve } from 'path';
import { OpenAPIV3 } from 'openapi-types';
import { toArrayTypes } from '../toArrayTypes';
import { TypesMap } from '../../types';
import { applyEslint } from '../../../__test__/utils/code';

const openApiMockResponse200 = readYaml(
  resolve(__dirname, '../../../mock/cases/response200.yaml'),
) as OpenAPIV3.Document;

const typesMap: TypesMap = {
  number: 'number',
  integer: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

const config = { typesMap };

describe('toArrayTypes mock/cases', () => {
  it('schemes in responses200.yaml', async () => {
    let code = toArrayTypes(openApiMockResponse200, config).join('\n');
    code = await applyEslint(code);

    expect(code).toMatchSnapshot();
  });
});
