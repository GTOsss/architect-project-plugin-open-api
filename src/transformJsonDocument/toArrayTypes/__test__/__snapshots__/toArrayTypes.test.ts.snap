// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toArrayTypes mock/cases schemes in responses200.yaml 1`] = `
"export type User = {
  id: number;
  email: string;
  name: string;
  roleEnumUnnamed: 'user_role' | 'admin_role' | 'guest_role';
  roleEnumReferenced: RoleEnum;
  tags: Array<string>;
  metadata: AnyObject;
  birthDate: string;
  score: number;
  isActive: boolean;
  address: Address;
};
export type Address = { street: string; city: string; country: string; zipCode: string };
export type RoleEnum = 'admin' | 'user' | 'guest';
"
`;
